src/Altinn.App.Core/Helpers/LazyRefreshCache.cs

```diff
@@ -0,0 +1,224 @@
+    /// </summary>
+    /// <param name="key">The cache key</param>
+    /// <param name="valueFactory">Factory function to create the value (if not in cache).</param>
+    /// <param name="lifetime">The lifetime of the cached value.</param>
+    /// <param name="postProcessCallback">Optional post processing callback (for metrics, etc)</param>
+    public async Task<TValue> GetOrCreate(
+        TKey key,
+        Func<Task<TValue>> valueFactory,
+        TimeSpan? lifetime = default,
+        Action<TValue?, CacheResultType>? postProcessCallback = default
+    )
+    {
+        return await GetOrCreate(
+            key: key,
+            valueFactory: valueFactory,
+            lifetimeFactory: _ => lifetime ?? DateTimeOffset.MaxValue - _timeProvider.GetUtcNow(),
+            postProcessCallback: postProcessCallback
+        );
+    }
+
+    private async Task<TValue> GetTaskFromTaskCache(
+        TKey key,
+        Func<Task<TValue>> valueFactory,
+        Func<TValue, TimeSpan> lifetimeFactory,
+        DateTimeOffset now,
+        Action<TValue?, CacheResultType>? postProcessCallback = default
+    )
+    {
+        return await _taskCache
+            .GetOrAdd(
+                key,
+                _ => new Lazy<Task<TValue>>(async () =>
+                {
+                    try
+                    {
+                        var valueTask = valueFactory();
+                        var value = await valueTask;
+                        var lifetime = lifetimeFactory(value);
+                        var expiry = now + lifetime;
+                        var refreshAt = expiry - _refetchBeforeExpiry;
+                        var cacheEntry = new CacheEntry(expiry, refreshAt, value);
+
+                        // Remove expired items in _valueCache
+                        foreach (var expiredItem in _valueCache.Where(x => x.Value.Expiry <= now))
+                        {
+                            _valueCache.TryRemove(expiredItem);
+                        }
+
+                        // Store the new value in _valueCache
+                        _valueCache.AddOrUpdate(key, cacheEntry, (_, _) => cacheEntry);
+
+                        // Too many items in _valueCache?
+                        if (_maxCacheEntries > 0 && _valueCache.Count > _maxCacheEntries)
+                        {
+                            var oldestEntries = GetOverflowKeys(_maxCacheEntries);
+                            foreach (var oldKey in oldestEntries)
+                            {
+                                _valueCache.TryRemove(oldKey);
+                            }
+                        }
+
+                        // Remove the Lazy Task from the _taskCache because it is no longer needed
+                        // and need to be gone when the cache is refreshed
+                        _taskCache.TryRemove(key, out var _);
+
+                        return value;
+                    }
+                    catch (Exception e)
```

Comment:
nit: No need to assign var if not to be used.`catch (Exception e)` ->`catch (Exception)`